### PR TITLE
[MRG] ADD split_naming arg to save to make BIDS-compatible files

### PR DESCRIPTION
**Pull request by teonbrooks**
Fixes https://github.com/mne-tools/mne-bids/issues/17

This adds `split_naming` arg to the save method to make BIDS-compatible files

### CODE REVIEW

File: mne/io/base.py [hunk 1]
```diff
@@ -2186,14 +2204,24 @@ def _write_raw(fname, raw, info, picks, fmt, data_type, reset_range, start,
...
+            end_block(fid, FIFF.FIFFB_REF)
+    else:
+        if part_idx > 1 and prev_fname is not None:
+            start_block(fid, FIFF.FIFFB_REF)
+            write_int(fid, FIFF.FIFF_REF_ROLE, FIFF.FIFFV_ROLE_PREV_FILE)
+            write_string(fid, FIFF.FIFF_REF_FILE_NAME, prev_fname)
+            if info['meas_id'] is not None:
+                write_id(fid, FIFF.FIFF_REF_FILE_ID, info['meas_id'])
+            write_int(fid, FIFF.FIFF_REF_FILE_NUM, part_idx - 2)
+            end_block(fid, FIFF.FIFFB_REF)
```
Comment (agramfort) [inline mne/io/base.py]: Please do this with less copy paste

File: mne/io/base.py [hunk 2]
```diff
@@ -1734,9 +1737,18 @@ def save(self, fname, picks=None, tmin=0, tmax=None, buffer_size_sec=None,
...
-                   start, stop, buffer_size, projector, drop_small_buffer,
-                   split_size, 0, None)
+        if split_naming == 'elekta':
+            _write_raw(fname, self, info, picks, fmt, data_type, reset_range,
+                       start, stop, buffer_size, projector, drop_small_buffer,
+                       split_size, split_naming, 0, None)
+        elif split_naming == 'bids':
+            _write_raw(fname, self, info, picks, fmt, data_type, reset_range,
+                       start, stop, buffer_size, projector, drop_small_buffer,
+                       split_size, split_naming, 1, None)
```
Comment (agramfort) [inline mne/io/base.py]: Please do this with less copy paste

File: mne/io/base.py [hunk 3]
```diff
@@ -1734,9 +1737,17 @@ def save(self, fname, picks=None, tmin=0, tmax=None, buffer_size_sec=None,
         buffer_size = self._get_buffer_size(buffer_size_sec)
 
         # write the raw file
+        if split_naming == 'elekta':
+            part_idx = 0
+        elif split_naming == 'bids':
+            part_idx = 1
+        else:
+            err = ("split_naming must be either 'elekta' or 'bids instead of'"
```
Comment (agramfort) [inline mne/io/base.py]: I think you miss an empty space after the "of"

File: mne/io/base.py [hunk 4]
```diff
@@ -2186,13 +2202,17 @@ def _write_raw(fname, raw, info, picks, fmt, data_type, reset_range, start,
         write_int(fid, FIFF.FIFF_FIRST_SAMPLE, first_samp)
 
     # previous file name and id
+    if split_naming == 'elekta':
+        data = part_idx - 1
+    else:
+        data = part_idx - 2
```
Comment (agramfort) [inline mne/io/base.py]: data is not a great name for an "int"

can it be a name based on part_idx ? part_idx_tag ?

Comment (teonbrooks) [reply mne/io/base.py]: i got the name data from the helper function ;) but i will use a more descriptive one

File: mne/io/base.py [hunk 5]
```diff
@@ -1734,9 +1737,17 @@ def save(self, fname, picks=None, tmin=0, tmax=None, buffer_size_sec=None,
...
 
         # write the raw file
+        if split_naming == 'elekta':
+            part_idx = 0
+        elif split_naming == 'bids':
+            part_idx = 1
+        else:
+            err = ("split_naming must be either 'elekta' or 'bids' instead "
+                   "of '{}'.".format(split_naming))
+            raise ValueError(err)
```
Comment (larsoner) [inline mne/io/base.py]: Put the string in the `ValueError` itself, no need for this splitting business. If you like the spacing better just do:
```
raise ValueError(
    "...")
```

File: mne/io/base.py [hunk 6]
```diff
@@ -1661,6 +1662,8 @@ def save(self, fname, picks=None, tmin=0, tmax=None, buffer_size_sec=None,
             .. note:: Due to FIFF file limitations, the maximum split
                       size is 2GB.
 
+        split_naming : string ('elekta' | 'bids')
+            Add the filename partition with the appropriate naming schema.
```
Comment (larsoner) [inline mne/io/base.py]:
```
    ... schema.

    .. versionadded:: 0.17
verbose: ...
```

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 7]
```diff
@@ -398,9 +398,23 @@ def test_split_files():
...
 
+    # check that the filenames match the intended pattern
+    assert op.exists(split_fname_elekta_part2)
+    # check that filenames are being formatted correctly for BIDS
+    raw_1.save(split_fname,
+               buffer_size_sec=1.0, split_size='10MB',
+               split_naming='bids',
+               overwrite=True)
+    assert op.exists(split_fname_bids_part1)
+    assert op.exists(split_fname_bids_part2)
```
Comment (larsoner) [inline mne/io/fiff/tests/test_raw_fiff.py]: should also be easy to test that `np.allclose` on the data when reading the `split_fname` (i.e., I/O round-trip)

File: mne/io/base.py [hunk 8]
```diff
@@ -1661,6 +1662,11 @@ def save(self, fname, picks=None, tmin=0, tmax=None, buffer_size_sec=None,
             .. note:: Due to FIFF file limitations, the maximum split
                       size is 2GB.
 
+        split_naming : string ('neuromag' | 'bids')
```
Comment (larsoner) [inline mne/io/base.py]: Sorry I missed this one, curly braces not `()` like `split_naming : {'neuromag', 'bids'}` https://numpydoc.readthedocs.io/en/latest/format.html

**Comment by teonbrooks** [comment]
hmm....the tests are passing locally and there isn't an explicit error message about why it is breaking on ci. does anyone have a chance to look at this?

**Comment by larsoner** [comment]
You have an un-caught warning. Catch it with `pytest.warns`

**Comment by larsoner** [comment]
(this should also not pass locally if you rebase on `master`)

**Comment by larsoner** [comment]
(our CIs turn un-caught warnings into errors nowadays)

**Comment by larsoner** [comment]
> are we going to keep the file naming warning? the bids nomenclature will throw this every time

Probably not but that will require a bit of work and convergence to fix. You can fix it here if you don't mind the delays. Otherwise, catch the warnings and we can merge

**Comment by larsoner** [comment]
line 430

https://travis-ci.org/mne-tools/mne-python/jobs/421731199#L2764

**Comment by teonbrooks** [comment]
> line 430

is there a warning on reading too?

also, is this the proper way to catch the filename warning: `with pytest.warns(RuntimeWarning)`

**Comment by larsoner** [comment]
Yes on read and write. Best to test locally so you don't have to wait so long to find out if it's working.

It's better to do `with pytest.warns(RuntimeWarning, match='some matching regex')` so a) it's clear what's being caught and b) you ensure you catch the right thing

**Comment by teonbrooks** [comment]
it's been a while, i'm a bit rusty at it all ;)

**Comment by teonbrooks** [comment]
@agramfort indeed. these days, js is all the rage ;) been learning react and js.

i noticed something that this PR is missing, it isn't doing an overwrite check on the bids filename. the filename isn't generated til the helper function `_write_raw` whereas typical fiiles are generated in the `save` method. just need to work that out first

**Comment by agramfort** [comment]
>
> @agramfort <https://github.com/agramfort> indeed. these days, js is all
> the rage ;) been learning react and js.
>
good. Want to do an EEG web raw data browser or a jupyter widget on a
sunday afternoon ? :)

**Comment by teonbrooks** [comment]
> good. Want to do an EEG web raw data browser or a jupyter widget on a sunday afternoon ? :)

with the [iodide](https://iodide.io/) project and [pyodide](https://github.com/iodide-project/pyodide) support, soon you will be able to embed `mne_browse_raw` in a webpage :) currently wip on scipy to webassembly and after that `mne` 🧠

**Comment by teonbrooks** [comment]
@larsoner or @agramfort, I think it is ready now

File: mne/io/fiff/tests/test_raw_fiff.py (hunk 7 above)
Comment (larsoner) [reply mne/io/fiff/tests/test_raw_fiff.py]: not addressed

Comment (larsoner) [reply mne/io/fiff/tests/test_raw_fiff.py]: ... oh wait, it is below!

File: setup.cfg [hunk 9]
```diff
@@ -23,13 +23,12 @@ filterwarnings =
...
     ignore:the matrix subclass:PendingDeprecationWarning
     ignore:numpy.dtype size changed:RuntimeWarning
+    ignore:module pycuda not found:RuntimeWarning
     ignore:.*HasTraits.trait_.*:DeprecationWarning
     ignore:.*takes no parameters:DeprecationWarning
     ignore:joblib not installed:RuntimeWarning
     ignore:Using a non-tuple sequence for multidimensional indexing:FutureWarning
     ignore:using a non-integer number instead of an integer will result in an error:DeprecationWarning
-    ignore:Importing from numpy.testing.decorators is deprecated, import from numpy.testing instead:DeprecationWarning
-    ignore:np.loads is deprecated, use pickle.loads instead:DeprecationWarning
```
Comment (larsoner) [inline setup.cfg]: This looks like rebasing errors

**Comment by agramfort** [comment]
besides @larsoner comment LGTM

**Comment by teonbrooks** [comment]
cool. all addressed

**Comment by larsoner** [comment]
Thanks @teonbrooks ! I should look forward to many more PRs from you, right? :)

**Comment by teonbrooks** [comment]
@larsoner the one really nagging PR idea I do have in mind is working on creating the `Digitization` class and finally addressing the confusion around `Montage` and `DigMontage`. i might get nerd-sniped into it soon enough ;)

**Comment by larsoner** [comment]
That would be amazing. Lots of users would use it and it would simplify our code a lot to have this unified class!

**Pull request merged**